src/transformers/models/whisper/modeling_whisper.py

```diff
@@ -430,50 +413,54 @@ def forward(
         # for the decoder
         is_cross_attention = key_value_states is not None
 
-        bsz, q_len, _ = hidden_states.size()
+        bsz, tgt_len, _ = hidden_states.size()
 
         # get query proj
-        query_states = self._reshape(self.q_proj(hidden_states), -1, bsz)
-        # get key, value proj
-        # `past_key_value[0].shape[2] == key_value_states.shape[1]`
-        # is checking that the `sequence_length` of the `past_key_value` is the same as
-        # the provided `key_value_states` to support prefix tuning
-        if (
-            is_cross_attention
-            and past_key_value is not None
-            and past_key_value[0].shape[2] == key_value_states.shape[1]
-        ):
-            # reuse k,v, cross_attentions
-            key_states = past_key_value[0].transpose(1, 2)
-            value_states = past_key_value[1].transpose(1, 2)
-        elif is_cross_attention:
-            # cross_attentions
-            key_states = self._reshape(self.k_proj(key_value_states), -1, bsz)
-            value_states = self._reshape(self.v_proj(key_value_states), -1, bsz)
-        elif past_key_value is not None:
-            # reuse k, v, self_attention
-            key_states = self._reshape(self.k_proj(hidden_states), -1, bsz)
-            value_states = self._reshape(self.v_proj(hidden_states), -1, bsz)
-            key_states = torch.cat([past_key_value[0].transpose(1, 2), key_states], dim=1)
-            value_states = torch.cat([past_key_value[1].transpose(1, 2), value_states], dim=1)
+        query_states = self._shape(self.q_proj(hidden_states), tgt_len, bsz)
+        past_key_value = getattr(self, "past_key_value", past_key_value)
+        if is_cross_attention:
+            # decoder cross-attention
+            if (
+                past_key_value is not None
+                and (isinstance(past_key_value, StaticCache) and not past_key_value.is_initialized[self.layer_idx])
+                or past_key_value.get_seq_length(self.layer_idx)
+            ):
+                # reuse k,v, cross_attentions
+                key_states = past_key_value.key_cache[self.layer_idx]
+                value_states = past_key_value.value_cache[self.layer_idx]
+            else:
+                # compute cross_attentions
+                key_states = self._shape(self.k_proj(key_value_states), -1, bsz)
+                value_states = self._shape(self.v_proj(key_value_states), -1, bsz)
+                if past_key_value is not None:
+                    # save all cross attention key/value_states to cache
+                    # further calls to cross_attention layer can then reuse all cross-attention
+                    cache_position = torch.arange(key_states.size(2), device=key_states.device)
+                    key_states, value_states = past_key_value.update(
+                        key_states, value_states, self.layer_idx, {"cache_position": cache_position}
+                    )
```

Comment:
since we are updating this, which was taken from bart, we can simplify it with what was done in UMt5:
```python
        is_cross_attention = encoder_hidden_states is not None
        batch_size, seq_length = hidden_states.shape[:2]

        # use encoder_hidden_states if cross attention
        current_states = encoder_hidden_states if encoder_hidden_states is not None else hidden_states
        # checking that the `sequence_length` of the `past_key_value` is the same as the he provided
        # `encoder_hidden_states` to support prefix tuning
        if is_cross_attention and past_key_value and past_key_value[0].shape[2] == current_states.shape[1]:
            # reuse k,v, cross_attentions
            key_states = past_key_value[0]
            value_states = past_key_value[1]
        else:
            key_states = self._shape(self.k(current_states))
            value_states = self._shape(self.v(current_states))
            if past_key_value is not None and not is_cross_attention:
                # reuse k, v, self_attention
                key_states = torch.cat([past_key_value[0], key_states], dim=2)
                value_states = torch.cat([past_key_value[1], value_states], dim=2)

```

Reply:
Resolved in https://github.com/huggingface/transformers/pull/31166/commits/2d917081c4a65de714105c1ba81b2f2b6a068652